chain/xrp/tx/tx.go

```
@@ -0,0 +1,229 @@
+	if xrpTx.LastLedgerSequence == 0 {
+		return []byte{}, errors.New("missing last ledger sequence")
+	}
+
+	if xrpTx.TxnSignature == "" {
+		return []byte{}, errors.New("missing transaction signature")
+	}
+
+	resultMapXRP := RenderToMap(*xrpTx)
+	resultMapWithAmount := make(map[string]interface{})
+
+	if xrpTx.Amount.IsString {
+		resultMapWithAmount = WithTokenAmount(resultMapXRP, xrpTx.Amount.StringValue)
+	} else {
+		resultMapWithAmount = WithTokenAmount(resultMapXRP, xrpTx.Amount.AmountValue)
+	}
+
+	encodeTx, err := binarycodec.Encode(resultMapWithAmount)
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to serialise serialised XRP transaction: %v", err)
+	}
+
+	encodeTxWithPrefix := TRANSACTION_HASH_PREFIX + string(encodeTx)
+
+	decodedBytes, err := hex.DecodeString(encodeTxWithPrefix)
+
+	hash := sha512.Sum512(decodedBytes)
+
+	firstHalf := hash[:32]
+	hashHex := hex.EncodeToString(firstHalf)
+	fmt.Println(hashHex)
+
+	return []byte(encodeTx), nil
+}
+
+func RenderToMap(xrpTx XRPTransaction) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["Account"] = string(xrpTx.Account)
+	result["Amount"] = xrpTx.Amount
+	result["Destination"] = string(xrpTx.Destination)
+	result["Fee"] = xrpTx.Fee
+	result["Flags"] = int(xrpTx.Flags)
+	result["LastLedgerSequence"] = int(xrpTx.LastLedgerSequence)
+	result["Sequence"] = int(xrpTx.Sequence)
+	result["SigningPubKey"] = xrpTx.SigningPubKey
+	result["TransactionType"] = string(xrpTx.TransactionType)
+	result["TxnSignature"] = xrpTx.TxnSignature
+
+	return result
+}
+
+func WithTokenAmount(fields map[string]interface{}, amount interface{}) map[string]interface{} {
+	switch v := amount.(type) {
+	case string:
+		fields["Amount"] = v
+	case *Amount:
+		fields["Amount"] = map[string]interface{}{
+			"currency": v.Currency,
+			"issuer":   v.Issuer,
+			"value":    v.Value,
+		}
+	case AmountBlockchain:
+		fields["Amount"] = map[string]interface{}{}
+	default:
+		fmt.Println("Invalid amount type")
```

should return an error